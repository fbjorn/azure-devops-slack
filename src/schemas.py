from datetime import datetime

from pydantic import BaseModel, Field


class Person(BaseModel):
    name: str = Field(..., alias="displayName")
    email: str = Field(..., alias="uniqueName")
    id: str


class Links(BaseModel):
    class _Link(BaseModel):
        href: str

    self: _Link
    repository: _Link
    threads: _Link | None


class Comment(BaseModel):
    content: str
    published_date: datetime = Field(..., alias="publishedDate")
    links: Links = Field(..., alias="_links")
    author: Person

    @property
    def thread_id(self) -> str | None:
        if not self.links.threads:
            return None
        threads_url = self.links.threads.href
        return threads_url.split("/")[-1]


class Repository(BaseModel):
    name: str
    url: str = Field(..., alias="webUrl")


class PullRequest(BaseModel):
    repository: Repository
    author: Person = Field(..., alias="createdBy")
    title: str
    description: str | None
    url: str
    reviewers: list[Person]
    id: str = Field(..., alias="pullRequestId")


class Resource(BaseModel):
    comment: Comment | None
    pr: PullRequest | None = Field(None, alias="pullRequest")
    created_by: Person | None = Field(None, alias="createdBy")
    reviewers: list[Person] | None


class Message(BaseModel):
    text: str
    html: str
    markdown: str


class DevOpsEvent(BaseModel):
    event_type: str = Field(..., alias="eventType")
    resource: Resource
    message: Message
    detailed_message: Message = Field(..., alias="detailedMessage")


class EventType:
    PR_COMMENT = "ms.vss-code.git-pullrequest-comment-event"
    PR_UPDATE = "git.pullrequest.updated"
    PR_CREATED = "git.pullrequest.created"
