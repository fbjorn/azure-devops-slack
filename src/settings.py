from pathlib import Path
from typing import Optional

from pydantic import BaseModel, BaseSettings, SecretStr


class UserConfig(BaseModel):
    name_filter: str | None
    emails: list[str]
    slack_id: str


class Conf(BaseSettings):
    API_KEY: Optional[SecretStr]
    SLACK_BOT_TOKEN: Optional[SecretStr]
    USERS: list[UserConfig] = []
    LOG_EVENTS: bool = False

    class Config:
        env_file = Path(__file__).parent.parent / ".env"


conf = Conf()
