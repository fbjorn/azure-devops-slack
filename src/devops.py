import requests
from requests.auth import HTTPBasicAuth

from src.schemas import APICommit
from src.settings import conf


class DevOpsClient:
    def __init__(self):
        self._auth = HTTPBasicAuth("", conf.DEVOPS_PAT.get_secret_value())

    def _fetch_url(self, url: str) -> dict:
        resp = requests.get(
            url,
            headers={"Content-type": "application/json"},
            auth=self._auth,
        )
        if resp.status_code != 200:
            resp.raise_for_status()

        return resp.json()

    def fetch_commit(self, url: str) -> APICommit | None:
        if not conf.DEVOPS_PAT:
            return None
        return APICommit(**self._fetch_url(url))
