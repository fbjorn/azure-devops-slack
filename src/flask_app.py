# This file is used in Docker only
from flask import Flask, request

from src.api import handle_api_event


def create_app():
    app = Flask("src.flask_app")

    @app.post("/")
    def handle_event():
        return handle_api_event(request)

    return app


application = create_app()
