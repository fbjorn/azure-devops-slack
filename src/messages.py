import re
from abc import abstractmethod

from pydantic import BaseModel

from src.schemas import DevOpsEvent
from src.utils import convert_links_to_slack_md


class SlackMessage(BaseModel):
    receiver: str
    evt: DevOpsEvent

    @abstractmethod
    def get_blocks(self) -> list:
        raise NotImplementedError

    def get_payload(self):
        return {"blocks": [self.get_blocks()]}


class PRCreated(SlackMessage):
    def get_blocks(self):
        text = convert_links_to_slack_md(self.evt.message.markdown)
        text += " and added you as a reviewer"
        return [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": text,
                },
            }
        ]


class CommentAdded(SlackMessage):
    def get_blocks(self) -> list:
        author = self.evt.resource.comment.author.name
        pr_title = self.evt.resource.pr.title
        repo = self.evt.resource.pr.repository.name
        pr = self.evt.resource.pr
        pr_url = f"{pr.repository.url}/pullrequest/{pr.id}"
        comment_url = f"{pr_url}?discussionId={self.evt.resource.comment.thread_id}"

        text = self.evt.resource.comment.content
        pics = re.findall(r"!\[.*?]\(.*?\)", text)
        for i, pic in enumerate(pics):
            text = text.replace(pic, "[SCREENSHOT]")

        return [
            {
                "type": "section",
                "text": {
                    "type": "mrkdwn",
                    "text": f"*{author} left a* <{comment_url}|comment>*:*",
                },
            },
            {
                "type": "section",
                "text": {"type": "mrkdwn", "text": f"{text}"},
            },
            {
                "type": "context",
                "elements": [{"type": "mrkdwn", "text": f"*{repo}* / {pr_title}"}],
            },
        ]


class PRStatusChanged(SlackMessage):
    def get_blocks(self) -> list:
        approved = "approved pull request" in self.evt.message.text
        waiting = "waiting for" in self.evt.message.text
        text = convert_links_to_slack_md(self.evt.message.markdown)
        if approved:
            text = f"âœ… {text}"
        elif waiting:
            text = f"ðŸš§ {text}"

        if waiting or approved:
            return [
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": text,
                    },
                },
            ]
