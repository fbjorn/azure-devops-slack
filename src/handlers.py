from functools import wraps
from typing import List

from src.messages import CommentAdded, PRCreated, PRStatusChanged, SlackMessage
from src.schemas import DevOpsEvent, EventType, Person
from src.settings import UserConfig, conf

EVENT_HANDLERS = {}


def find_user(person: Person) -> UserConfig | None:
    for user in conf.USERS:
        for user_email in user.emails:
            if user_email == person.email:
                return user
    for user in conf.USERS:
        if user.name_filter in person.name.lower():
            return user
    return None


def event_handler(event: EventType):
    def handle_event(fn):
        @wraps(fn)
        def _handle_event(*args, **kwargs):
            result = fn(*args, **kwargs)
            return result if result is not None else []

        EVENT_HANDLERS[event] = _handle_event
        return _handle_event

    return handle_event


def compose_messages_from_event(evt: DevOpsEvent) -> List[SlackMessage]:
    return EVENT_HANDLERS[evt.event_type](evt)


@event_handler(EventType.PR_COMMENT)
def parse_comment_added(evt: DevOpsEvent) -> List[SlackMessage]:
    pr_author_id = evt.resource.pr.author.id
    comment_author_id = evt.resource.comment.author.id

    if pr_author_id != comment_author_id:
        if author := find_user(evt.resource.pr.author):
            return [CommentAdded(receiver=author.slack_id, evt=evt)]
    else:
        # TODO: Author left the comment. Need to check if it's a reply
        #   to another user and notify that user
        return []


@event_handler(EventType.PR_CREATED)
def parse_pr_created_event(evt: DevOpsEvent) -> List[SlackMessage] | None:
    messages = []
    for reviewer in evt.resource.reviewers:
        if user := find_user(reviewer):
            messages.append(PRCreated(receiver=user.slack_id, evt=evt))
    return messages


@event_handler(EventType.PR_UPDATE)
def parse_pr_updated_event(evt: DevOpsEvent) -> List[SlackMessage] | None:
    messages = []
    approved = "approved pull request"
    waits = "is waiting for the author"
    if approved in evt.message.text or waits in evt.message.text:
        if user := find_user(evt.resource.created_by):
            messages.append(PRStatusChanged(receiver=user.slack_id, evt=evt))
    # else:
    #     for reviewer in evt.resource.reviewers:
    #         if user := find_user(reviewer):
    #             messages.append(PRCreated(receiver=user.slack_id, evt=evt))
    return messages
