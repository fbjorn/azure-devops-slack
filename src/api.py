import secrets

from flask import Request, Response

from src.handlers import compose_messages_from_event
from src.schemas import DevOpsEvent
from src.settings import conf
from src.slack import send_direct_message


def handle_api_event(request: Request):
    api_key = request.headers.get("X-API-KEY")
    if conf.API_KEY and not secrets.compare_digest(
        conf.API_KEY.get_secret_value(), api_key
    ):
        return Response(status=401)

    payload = request.get_json(silent=True) or {}
    if conf.LOG_EVENTS:
        print(payload)

    event = DevOpsEvent(**payload)
    messages = compose_messages_from_event(event)
    for message in messages:
        send_direct_message(message)

    return "OK"
