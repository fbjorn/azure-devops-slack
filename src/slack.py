import requests

from src.messages import SlackMessage
from src.settings import conf

POST_MESSAGE_URL = "https://slack.com/api/chat.postMessage"


def send_direct_message(message: SlackMessage):
    if not conf.SLACK_BOT_TOKEN:
        raise ValueError("Slack bot token is missing")

    resp = requests.post(
        POST_MESSAGE_URL,
        headers={
            "Accept": "Application/JSON",
            "Authorization": f"Bearer {conf.SLACK_BOT_TOKEN.get_secret_value()}",
        },
        json={
            "channel": message.receiver,
            "blocks": message.get_blocks(),
        },
    )
    if resp.status_code != 200:
        print("Error while sending Slack message")
