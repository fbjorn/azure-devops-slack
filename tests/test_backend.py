from unittest.mock import MagicMock

import pytest
from pydantic import SecretStr

from src.flask_app import create_app
from src.messages import SlackMessage
from tests.helpers import read_payload_as_json


@pytest.fixture()
def client():
    app = create_app()
    app.config.update({"TESTING": True})
    with app.test_client() as test_client:
        yield test_client


@pytest.fixture
def random_event():
    return read_payload_as_json("comment_added")


@pytest.fixture
def slack(monkeypatch):
    mocked_slack = MagicMock()
    monkeypatch.setattr("src.api.send_direct_message", mocked_slack)
    return mocked_slack


def test_request_not_authorized(client, monkeypatch):
    monkeypatch.setattr("src.settings.conf.API_KEY", SecretStr("foo"))

    resp = client.post("/")
    assert resp.status_code == 401, "X-API-KEY is missing"

    resp = client.post("/", headers={"X-API-KEY": "bar"})
    assert resp.status_code == 401, "X-API-KEY is not correct"


def test_request_is_parsed(client, monkeypatch, random_event, slack, snapshot):
    monkeypatch.setattr("src.settings.conf.API_KEY", SecretStr("foo"))

    resp = client.post("/", headers={"X-API-KEY": "foo"}, json=random_event)
    assert resp.status_code == 200

    msg: SlackMessage = slack.call_args_list[0][0][0]
    assert msg.receiver == "jack-slack-id"
    assert msg.get_blocks() == snapshot


def test_request_is_parsed_without_api_key(
    client, monkeypatch, random_event, slack, snapshot
):
    monkeypatch.setattr("src.settings.conf.API_KEY", None)
    resp = client.post("/", json=random_event)
    assert resp.status_code == 200

    msg: SlackMessage = slack.call_args_list[0][0][0]
    assert msg.receiver == "jack-slack-id"
    assert msg.get_blocks() == snapshot
