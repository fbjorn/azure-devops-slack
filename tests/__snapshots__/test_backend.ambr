# serializer version: 1
# name: test_request_is_parsed
  list([
    dict({
      'text': dict({
        'text': '*Jane Doe left a* <https://dev.azure.com/MyOrg/MyProject/_git/my-repo/pullrequest/3369?discussionId=26855|comment>*:*',
        'type': 'mrkdwn',
      }),
      'type': 'section',
    }),
    dict({
      'text': dict({
        'text': 'Another comment on your request',
        'type': 'mrkdwn',
      }),
      'type': 'section',
    }),
    dict({
      'elements': list([
        dict({
          'text': '*my-repo* / Fix README',
          'type': 'mrkdwn',
        }),
      ]),
      'type': 'context',
    }),
  ])
# ---
# name: test_request_is_parsed_without_api_key
  list([
    dict({
      'text': dict({
        'text': '*Jane Doe left a* <https://dev.azure.com/MyOrg/MyProject/_git/my-repo/pullrequest/3369?discussionId=26855|comment>*:*',
        'type': 'mrkdwn',
      }),
      'type': 'section',
    }),
    dict({
      'text': dict({
        'text': 'Another comment on your request',
        'type': 'mrkdwn',
      }),
      'type': 'section',
    }),
    dict({
      'elements': list([
        dict({
          'text': '*my-repo* / Fix README',
          'type': 'mrkdwn',
        }),
      ]),
      'type': 'context',
    }),
  ])
# ---
